Content.Shared/Doors/Systems/SharedDoorSystem.Airlock.cs

```diff
@@ -0,0 +1,189 @@
+
+namespace Content.Shared.Doors.Systems;
+
+public abstract partial class SharedDoorSystem
+{
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] protected readonly SharedDoorSystem DoorSystem = default!;
+    [Dependency] private readonly SharedWiresSystem _wiresSystem = default!;
+
+    private void InitializeAirlock()
+    {
+        SubscribeLocalEvent<AirlockComponent, BeforeDoorClosedEvent>(OnBeforeDoorClosed);
+        SubscribeLocalEvent<AirlockComponent, DoorStateChangedEvent>(OnStateChanged);
+        SubscribeLocalEvent<AirlockComponent, DoorBoltsChangedEvent>(OnBoltsChanged);
+        SubscribeLocalEvent<AirlockComponent, BeforeDoorOpenedEvent>(OnBeforeDoorOpened);
+        SubscribeLocalEvent<AirlockComponent, BeforeDoorDeniedEvent>(OnBeforeDoorDenied);
+        SubscribeLocalEvent<AirlockComponent, GetPryTimeModifierEvent>(OnGetPryMod);
+        SubscribeLocalEvent<AirlockComponent, BeforePryEvent>(OnBeforePry);
+    }
+
+    private void OnBeforeDoorClosed(Entity<AirlockComponent> airlock, ref BeforeDoorClosedEvent args)
+    {
+        if (args.Cancelled)
+            return;
+
+        if (!airlock.Comp.Safety)
+            args.PerformCollisionCheck = false;
+
+        // only block based on bolts / power status when initially closing the door, not when its already
+        // mid-transition. Particularly relevant for when the door was pried-closed with a crowbar, which bypasses
+        // the initial power-check.
+
+        if (!TryComp(airlock, out DoorComponent? door)
+            || door.State is DoorState.Opening or DoorState.Closing
+            || CanChangeState(airlock,
+                door.State is DoorState.AttemptingCloseByPrying or DoorState.AttemptingOpenByPrying))
+            return;
+
+        args.Cancel();
+    }
+
+    private void OnStateChanged(Entity<AirlockComponent> airlock, ref DoorStateChangedEvent args)
+    {
+        // This is here, so we don't accidentally bulldoze state values and mis-predict.
+        if (_timing.ApplyingState)
+            return;
+
+        // Only show the maintenance panel if the airlock is closed
+        if (TryComp<WiresPanelComponent>(airlock, out var wiresPanel))
+        {
+            _wiresSystem.ChangePanelVisibility(airlock,
+                wiresPanel,
+                airlock.Comp.OpenPanelVisible || args.State != DoorState.Open);
+        }
+
+        // If the door is closed, we should look if the bolt was locked while closing
+        UpdateAutoClose(airlock);
+
+        // Make sure the airlock auto closes again next time it is opened
+        if (args.State != DoorState.Closed)
+            return;
+
+        airlock.Comp.AutoClose = true;
+        Dirty(airlock);
+    }
+
+    private void OnBoltsChanged(Entity<AirlockComponent> airlock, ref DoorBoltsChangedEvent args)
+    {
+        // If unbolted, reset the auto close timer
+        if (args.BoltsDown)
+            return;
+
+        UpdateAutoClose(airlock);
```

Comment:
```suggestion
        // If unbolted, reset the auto close timer
        if (!args.BoltsDown)
            UpdateAutoClose(airlock);
```
Confusing that the "default" behavior is particular to a state.